src/Sylius/Bundle/ApiBundle/EventListener/ApiCartBlamerListener.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\EventListener;
+
+use Doctrine\Persistence\ObjectManager;
+use Sylius\Bundle\ApiBundle\SectionResolver\ShopApiOrdersSubSection;
+use Sylius\Bundle\CoreBundle\SectionResolver\SectionProviderInterface;
+use Sylius\Bundle\UserBundle\Event\UserEvent;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\Order\Context\CartContextInterface;
+use Sylius\Component\Order\Context\CartNotFoundException;
+use Sylius\Component\Resource\Exception\UnexpectedTypeException;
+use Symfony\Component\Security\Http\Event\InteractiveLoginEvent;
+
+final class ApiCartBlamerListener
+{
+    /** @var ObjectManager */
+    private $cartManager;
+
+    /** @var CartContextInterface */
+    private $cartContext;
+
+    /** @var SectionProviderInterface */
+    private $uriBasedSectionContext;
+
+    public function __construct(
+        ObjectManager $cartManager,
+        CartContextInterface $cartContext,
+        SectionProviderInterface $uriBasedSectionContext
+    ) {
+        $this->cartManager = $cartManager;
+        $this->cartContext = $cartContext;
+        $this->uriBasedSectionContext = $uriBasedSectionContext;
+    }
+
+    public function onImplicitLogin(UserEvent $userEvent): void
+    {
+        if (!$this->uriBasedSectionContext->getSection() instanceof ShopApiOrdersSubSection) {
+            return;
+        }
+
+        $user = $userEvent->getUser();
+        if (!$user instanceof ShopUserInterface) {
+            return;
+        }
+
+        $this->blame($user);
+    }
+
+    public function onInteractiveLogin(InteractiveLoginEvent $interactiveLoginEvent): void
+    {
+        if (!$this->uriBasedSectionContext->getSection() instanceof ShopApiOrdersSubSection) {
+            return;
+        }
+
+        $user = $interactiveLoginEvent->getAuthenticationToken()->getUser();
+        if (!$user instanceof ShopUserInterface) {
+            return;
+        }
+
+        $this->blame($user);
+    }
+
+    private function blame(ShopUserInterface $user): void
+    {
+        $cart = $this->getCart();
+        if (null === $cart) {
```

Comment:
in addition, we should also check if the cart has a customer already assigned, and if so early return as well